### PR TITLE
add errorsink.d, an abstract interface to error messages

### PR DESCRIPTION
**Pull request by WalterBright**
I've made considerable progress into making the lexer standalone, and the lexer/parser standalone. The problem is its reliance on globals.d and errors.d, which pull in a bunch of other stuff not relevant to lexing/parsing. The code in errors.d is of rather stupefying complexity, which is completely irrelevant to the lexer/parser. This complexity also makes it difficult to create a standalone lexer/parser.

Enter the abstract interface for error messages, ErrorSink, in errorsink.d. It's only got the functions in it that the lexer/parser actually need. The lexer/parser should not need to know anything about color formatting, highlighting, gagging, or where the messages are sent to.

The constructors for the lexer/parser are passed an instance of ErrorSink.

This isn't perfect, there's still an ugly wart in it that relies on getenv. I'll tackle that problem after this one is incorporated.

Sorry about the high line count, although nearly all of it is just changing the call to `error()`. The substantive changes are pretty small.

The more we can reduce all these "everybody imports everybody else" the more tractable the code base will be.

### CODE REVIEW

**Comment by ibuclaw** [comment]
Errrrrrrrrrrrr @jacob-carlborg ?

**Comment by ibuclaw** [comment]
#9350
#9484
#9480
#9507
#10072
#10711

**Comment by jacob-carlborg** [comment]
🤦

File: compiler/src/dmd/lexer.d
Comment (RazvanN7) [inline compiler/src/dmd/lexer.d:25]: Shouldn't the use of errorsink get rid of this import?

**Comment by thewilsonator** [comment]
Yay, lets add even more files before packaging dmd properly.

**Comment by ibuclaw** [comment]
> Yay, lets add even more files before packaging dmd properly.

Worse than that, this PR literally [reinvents the wheel](https://github.com/dlang/dmd/blob/aa19d8a145afeb627a1136bf58f958998cd0fcec/compiler/src/dmd/frontend.d#L471-L683)... why?

File: compiler/src/dmd/errorsink.d
Comment (thewilsonator) [inline compiler/src/dmd/errorsink.d:19]: Reuse [existing wheels](https://github.com/dlang/dmd/blob/aa19d8a145afeb627a1136bf58f958998cd0fcec/compiler/src/dmd/frontend.d#L471-L683) or explain why they are insufficient. 

Such deficiencies might include `dmd.frontend` imports `std`.

Note also that `dmd.frontend` is a stable externally available API and so at worst this PR should move the definition from there to here, and have a publicly imported alias to the new place.

**Review by thewilsonator — DISMISSED** [review]

**Review by ljmf00 — DISMISSED** [review]
Please explain why we defer the creation of a proper error handling mechanism on DMD and also why not use the available interfaces exposed to be used on diagnostics?

**Comment by WalterBright** [comment]
I'm mystified by the unittest runner:
```
./build unittest
(DC) LEXER-unittest
(DC) DMD-unittest
(RUN) DMD-UNITTEST
Success
```
However, the unittests were not run, as I put a printf in them. So looking at the -v output, I see it is linking in:
```
../generated/linux/release/64/lexer-unittest.o ../generated/linux/release/64/backend-unittest.o ../generated/linux/release/64/common-unittest.o
```
What? The unittest feature in D is designed so you compile with -unittest, run the program, and the unittests are run. Where are all these extra .o files for? And why won't the unittests run when I run the generated dmd? What happened to our simple way of doing unittests?

**Comment by WalterBright** [comment]
> Why?

The best way to make understandable code is to have all the inputs and outputs at the entry point to the function. John Carmack explains this far better than I in http://www.sevangelatos.com/john-carmack-on/

What I've been doing over several previous PRs is just that - giving Lexer everything it needs via the call to its constructor, rather than "side loading" things through global variables, like `global`. This PR addresses a remaining problem - where do error messages go? The Lexer should not care what happens to the error messages, it just needs a place to send them. This is abstracted into ErrorSink, which is just what it says, it's a sink for error messages. Taking a look at the abstract class, it's utterly trivial and can be understood completely in 30 seconds. A couple handy implementations of it are there as well - ErrorSinkNull which just does nothing, and ErrorSinkStderr which just writes the messages to stderr. Both are trivial.

Contrast this with dmd.errors. It's a bewildering panoply of functions. There's a hook in it, `diagnosticHandler`, which snatches away control. But the user still has to understand all that stuff in errors.d, and it still all gets pulled into the executable, along with all its dependencies. It's full of global state, exactly what is undesirable in encapsulating functionality.

This PR does not break diagnosticHandler, or change the existing code that interfaces with it, other than adding a parameter to the constructor of Lexer.

Everything works but the unittests. So far, I have not been able to debug that because I have not managed to make heads or tails of how the unittests are run. Something about unittests in dmd have gone completely off the rails into a swamp.

**Comment by WalterBright** [comment]
The following lines are in lexer.d:
```
if (!errorSink) { debug fprintf(stderr, "file %s\n", file.ptr); }
if (!errorSink) { debug printf("file %s\n", file.ptr); }
        assert(errorSink);
```
and the log file result is:
```
1) test CRLF and mixed line endings

core.exception.AssertError@src\dmd\lexer.d(133): Assertion failure
```
Why does neither printf show up in the log? This makes things very difficult to debug.

File: compiler/src/dmd/errorsink.d
Comment (WalterBright) [reply compiler/src/dmd/errorsink.d:19]: > Reuse existing wheels or explain why they are insufficient.

This line illustrates the insufficiency:
```
Returns: false if the message should also be printed to stderr, true otherwise
```
This is the wrong abstraction. The user of Lexer should be able to send the errors anywhere, baking into the API "stderr" or not is insufficent. `ErrorSink` puts no constraint on where the messages should go. They can even go nowhere, as `ErrorSinkNull` demonstrates.

It's other insufficiency is it is centered around a pointer to a function which overrides the default behavior. This is a primitive form of OOP that fell out of favor in the 1980s. The abstract class, for which the user adds derived classes with specific behaviors, is the more modern approach. The abstract class designer does not have to anticipate what the implementation classes will be doing.

**Comment by WalterBright** [comment]
> Worse than that, this PR literally reinvents the wheel ... why?

dmd.frontend provides an implementation, not an abstract interface. An abstract interface is much more adaptable. dmd.errorsink is an example of an abstract interface.

You're right that all those imports of `std`, like `std.algorithm`, are a deficiency. The compiler should be free of dependency on Phobos working.

dmd.frontend needs a modern redesign, but that is way beyond the scope of this PR for the time being.

Oddly enough, I am listed as the author of frontend.d. I didn't write it.

File: compiler/src/dmd/lexer.d
Comment (WalterBright) [reply compiler/src/dmd/lexer.d:25]: Yes, it should. But there are a couple other cases where it is referenced directly that I will fix in a subsequent PR.

File: compiler/src/dmd/errorsink.d
Comment (WalterBright) [reply compiler/src/dmd/errorsink.d:19]: I don't want to touch dmd.frontend in this PR. As I mentioned to @ibuclaw it needs a significant overhaul.

File: compiler/src/dmd/lexer.d
Comment (WalterBright) [reply compiler/src/dmd/lexer.d:2659]: They're just convenience functions. I suppose they could be marked as protected, but not in this PR.

File: compiler/src/dmd/errorsink.d
Comment (thewilsonator) [reply compiler/src/dmd/errorsink.d:19]: Then I consider this PR to require approval from @jacob-carlborg  before merging.

> don't want to touch dmd.frontend in this PR

I do not trust you will ever get around to this. And this adds yet more files without packaging things properly and I trust even less that you will ever do that.

> They can even go nowhere, as ErrorSinkNull demonstrates.

So can `DiagnositReporter`, `return true;`.

> As I mentioned to @ibuclaw it needs a significant overhaul.

Where? why? (Hint, no it doesn't).

**Review by thewilsonator — DISMISSED** [review]

**Comment by jacob-carlborg** [comment]
> Then I consider this PR to require approval from @jacob-carlborg before merging.

I'm not doing any code reviews. This PR is the perfect example of why I'm not contributing anymore.

File: compiler/src/dmd/dmodule.d
Comment (Geod24) [inline compiler/src/dmd/dmodule.d:764]: Why is global assigned to here ?

File: compiler/src/dmd/dmodule.d
Comment (Geod24) [inline compiler/src/dmd/dmodule.d:1389]: Unrelated change, should be its own commit

File: compiler/src/dmd/doc.d
Comment (Geod24) [inline compiler/src/dmd/doc.d:38]: Why do we add this import if we're just passing a variable without ever referencing its type ?

File: compiler/src/dmd/dsymbolsem.d
Comment (Geod24) [inline compiler/src/dmd/dsymbolsem.d:42]: Ditto import

File: compiler/src/dmd/dtemplate.d
Comment (Geod24) [inline compiler/src/dmd/dtemplate.d:7564]: Unrelated, own commit

File: compiler/src/dmd/errors.d
Comment (Geod24) [inline compiler/src/dmd/errors.d:31]: I don't think that's the right abstraction either. What about messages issued by `pragma(msg)` ?

File: compiler/src/dmd/expressionsem.d
Comment (Geod24) [inline compiler/src/dmd/expressionsem.d:44]: Useless import

File: compiler/src/dmd/globals.d
Comment (Geod24) [inline compiler/src/dmd/globals.d:353]: This should be moved out of `globals` and into main to reduce circular dependencies

File: compiler/src/dmd/iasmgcc.d
Comment (Geod24) [inline compiler/src/dmd/iasmgcc.d:389]: Likewise, this pattern is not good

File: compiler/src/dmd/mars.d
Comment (Geod24) [inline compiler/src/dmd/mars.d:43]: Useless import

File: compiler/test/unit/lexer/location_offset.d
Comment (Geod24) [inline compiler/test/unit/lexer/location_offset.d:20]: I think this should just be a function at this point, instead of copy-pasted code

**Review by Geod24 — DISMISSED** [review]
- Lots of useless imports;
- The lazy instantiation of globals in the middle of semantic is a big no;
- The two classes you provide, `ErrorSinkNull` and `ErrorSinkStderr`, are completely unused, something you have repeatedly complained about;

Instead of re-doing what other people have been wanting for years, perhaps we should just look into how/what @jacob-carlborg did, with `DiagnosticReporter`, which was also introduced gradually (starting with the lexer).

File: compiler/test/unit/lexer/location_offset.d
Comment (WalterBright) [reply compiler/test/unit/lexer/location_offset.d:20]: It's a class because it consists of several functions, not one.

File: compiler/src/dmd/mars.d
Comment (WalterBright) [reply compiler/src/dmd/mars.d:43]: Yes, that's leftover from a previous incarnation of this PR.

File: compiler/src/dmd/iasmgcc.d
Comment (WalterBright) [reply compiler/src/dmd/iasmgcc.d:389]: You're right, that's left over from a previous incarnation of this PR. I had a fair amount of trouble figuring out where the unittests were being initialized.

File: compiler/src/dmd/expressionsem.d
Comment (WalterBright) [reply compiler/src/dmd/expressionsem.d:44]: yes, leftover from a previous incarnation of this PR

File: compiler/src/dmd/dtemplate.d
Comment (WalterBright) [reply compiler/src/dmd/dtemplate.d:7564]: That came about because of my checkin process converts all tabs to spaces, and apparently there were tabs on that line. Being a bit of trivia, there's no issue here.

File: compiler/src/dmd/doc.d
Comment (WalterBright) [reply compiler/src/dmd/doc.d:38]: removed

File: compiler/src/dmd/dmodule.d
Comment (WalterBright) [reply compiler/src/dmd/dmodule.d:1389]: Another case of tabs being converted to spaces by my checkin procedure. Not worth another PR.

File: compiler/test/unit/lexer/location_offset.d
Comment (Geod24) [reply compiler/test/unit/lexer/location_offset.d:20]: I meant that at some point, we should realize that:
```D
scope lexer = new TestLexer(code);
// Or:
scope lexer = makeTestLexer(code);
```

is the way to go.

**Comment by WalterBright** [comment]
> Lots of useless imports;

Removed. Explained in the comments.

> The lazy instantiation of globals in the middle of semantic is a big no;

Removed. Explained in the comments.

> The two classes you provide, ErrorSinkNull and ErrorSinkStderr, are completely unused, something you have repeatedly complained about;

ErrorSinkNull I plan to use in errors.d to replace the use of:
```
auto gaggedErrorsSave = global.startGagging();
```
in errors.d, which is the perfect use case for it.

I plan to use ErrorSinkError in the unit tests for Lexer.

Both are also an illustration for anyone needing their own custom error message sink.

> Instead of re-doing what other people have been wanting for years, perhaps we should just look into how/what @jacob-carlborg did, with DiagnosticReporter, which was also introduced gradually (starting with the lexer).

ErrorSink is a better design. It's hard to think of anything simpler - if you can, I want to hear about it. Note also that the only import ErrorSink needs is dmd.location. I'd actually like to get it to zero imports. In general, using sinks is an excellent approach to eliminating arbitrary dependencies.

File: compiler/src/dmd/iasmgcc.d
Comment (WalterBright) [reply compiler/src/dmd/iasmgcc.d:389]: Unfortunately, this code doesn't call globals._init, and so global.errorSink is not initialized. So I have to initialize it here.

File: compiler/test/unit/lexer/location_offset.d
Comment (WalterBright) [reply compiler/test/unit/lexer/location_offset.d:20]: That's worth considering for a future PR. We already use both techniques in dmd.

**Comment by WalterBright** [comment]
This serves our goals of reducing technical debt and making the lexer/parser more amenable to being standalone by reducing its dependencies on other modules. It's ready to merge. @RazvanN7 @dkorpel

**Review by dkorpel — DISMISSED** [review]
Creating a copy of what already exists does not reduce technical debt. If there's deficiencies in `DiagnosticReporter`, why not fix that instead?

**Comment by WalterBright** [comment]
> Creating a copy of what already exists does not reduce technical debt.

To reiterate:
1. it results in lexer.d having a dependency on dmd.console.
2. if a diagnosticreporter is used, the code for the non-use case remains in the program, unused
3. I quote: 
```
 * The diagnostic handler.
 * If non-null it will be called for every diagnostic message issued by the compiler.
 * If it returns false, the message will be printed to stderr as usual.
```
it should not leave a builtin dependency on stderr. It should not have default behavior at all.
4. It uses an old-fashioned pointer-to-handler OOP design that was supplanted in 1990 (aka a "hook" interface)
5. abstract interfaces are a better design
6. It still leaves a dependency on the gagging global which is an execrable design practice
7. It is much more complicated than necessary
8. anyone can understand ErrorSink in 30 seconds
9. GUI programs don't have a console or stderr
10. it enables easy mock sinks, which are critical for unittesting
11. it "side loads" into Lexer, rather than going through the function parameter list

This PR does not delete diagnostichandler because that would break existing code. I wished to keep this PR as a refactoring. Migrating from diagnostichandler to the abstract interface can be done later.

File: compiler/src/dmd/lexer.d
Comment (WalterBright) [reply compiler/src/dmd/lexer.d:551]: It's problematic because the variadic argument list is clumsy to just forward things along.

Comment (dkorpel) [reply compiler/src/dmd/lexer.d:551]: It's not:
```
    final void error(T...)(const(char)* format, T args)
    {
        error(token.loc, format, args);
    }

    final void error(T...)(const ref Loc loc, const(char)* format, T args)
    {
        error(loc, format, args);
    }
```
What's really clumsy is an awkward mix of calls to `error` and `eSink.error` based on the number of arguments.

Comment (WalterBright) [reply compiler/src/dmd/lexer.d:551]: You're right. I didn't think of the template solution.

**Review by dkorpel — APPROVED** [review]

File: compiler/src/dmd/errorsink.d
Comment (maxhaton) [inline compiler/src/dmd/errorsink.d:28]: `warningSupplemental`?

File: compiler/src/dmd/errorsink.d
Comment (maxhaton) [inline compiler/src/dmd/errorsink.d:33]: `message` should also be included.

**Comment by maxhaton** [comment]
Is this any less dependant on gagging?

File: compiler/src/dmd/errors.d
Comment (maxhaton) [reply compiler/src/dmd/errors.d:31]:
```
bool pragmaMsgSemantic(Loc loc, Scope* sc, Expressions* args)
{
    if (!args)
        return true;
    foreach (arg; *args)
    {
        sc = sc.startCTFE();
        auto e = arg.expressionSemantic(sc);
        e = resolveProperties(sc, e);
        sc = sc.endCTFE();

        // pragma(msg) is allowed to contain types as well as expressions
        e = ctfeInterpretForPragmaMsg(e);
        if (e.op == EXP.error)
        {
            errorSupplemental(loc, "while evaluating `pragma(msg, %s)`", arg.toChars());
            return false;
        }
        if (auto se = e.toStringExp())
        {
            const slice = se.toUTF8(sc).peekString();
            fprintf(stderr, "%.*s", cast(int)slice.length, slice.ptr);
        }
        else
            fprintf(stderr, "%s", e.toChars());
    }
    fprintf(stderr, "\n");
    return true;
}
```
impl should really be using `message`, but anyway motivates the need for another category of output in this type of interface.

File: compiler/src/dmd/errorsink.d
Comment (WalterBright) [reply compiler/src/dmd/errorsink.d:28]: I only added the ones used.

File: compiler/src/dmd/errorsink.d
Comment (WalterBright) [reply compiler/src/dmd/errorsink.d:33]: I only added the ones used. Others can be added as necessary.

**Comment by WalterBright** [comment]
> Is this any less dependant on gagging?

Currently, no. But as I remarked upthread, there's an obvious case in colorHightlightCode() for using ErrorSinkNull instead of gagging. Eventually, I'd like to see all use of gagging replaced by ErrorSinkNull.

File: compiler/src/dmd/iasmgcc.d
Comment (ibuclaw) [reply compiler/src/dmd/iasmgcc.d:389]: Having everything in global is not ideal either.

File: compiler/src/dmd/iasmgcc.d
Comment (ibuclaw) [inline compiler/src/dmd/iasmgcc.d:76]: This just makes iasmgcc ugly to look at for no obvious benefit.

File: compiler/src/dmd/iasmgcc.d
Comment (ibuclaw) [inline compiler/src/dmd/iasmgcc.d:413]: At this point, might as well make it a default param instead, there are dozens of repetitions of this pattern of change.

File: compiler/src/dmd/errors.d
Comment (ibuclaw) [inline compiler/src/dmd/errors.d:43]: Dangerously close to hijacking gcc error diagnostics here. Or am I expected to just fork out the implementation to the C++ side of the compiler?

File: compiler/src/dmd/errorsink.d
Comment (ntrel) [inline compiler/src/dmd/errorsink.d:69]: Shouldn't `loc` come before "Error:" etc?

**Pull request merged**
